### ISSUE TITLE
Create infrastructure to sort posts by type

### ISSUE THREAD

**Issue by TJScalzo** [tholman/inspiring-online#162]
Now that we've updated the number of types there are for posts (#133) I think it'd be nice if we used those types to sort the posts. It was discussed a little bit [here](https://github.com/tholman/inspiring-online/issues/133#issuecomment-192291643) but this can be a place to discuss it more in detail.

**Comment by TJScalzo** [comment]
I'm not sure which way I should go with this. It'd be very easy to create an html page exactly like the homepage for every category that are then linked to on a main category page. Should I do that so that we at least have a simple base to improve on or should I try and do something I haven't done before. I'm going to start simple and do what I explained above. If someone wants to work out a single page that generates/sorts the posts in an easier, more extensible way go ahead.

**Comment by TJScalzo** [comment]
Okay, I've managed to make pages that sort the posts by their categories. I'm now working on an interesting way to display all of the categories.

**Comment by TJScalzo** [comment]
I have done a lot. Every post now has a label that says what type it is that links to all of the posts of that type. Everything is pretty much ready to go. The only thing that I need to do now is make a page that lists all of the types. 
The trouble I'm having with this is that I'm not sure how they should be shown. As buttons? As images of what is inside that category? I'm stuck and I need a little bit of help. I'll push my code up to my fork so everyone can see what I've done. I won't make a pull request yet because there's no real category page and so things would be a little confusing.

**Comment by TJScalzo** [comment]
I've pushed my changes. Everyone can find them [here](https://github.com/TJScalzo/inspiring-online).
#### Other things I need help/opinions on:
- Placement of links to categories
- Placement of titles of categories
- Colors for each category
- Word links vs. button links
- Neatening code (I'm sure there's a simpler way to do this)

**Comment by JohJakob** [comment]
@TJScalzo Wow, you did a good job! I would make the category link in posts a little bit smaller and not italic. Maybe, we can create some cool banner-like links like [Product Hunt's collection banners](https://www.producthunt.com/collections), but with some nice GIFs as the background images matching the color of its category link.

**Comment by TJScalzo** [comment]
Thabk you @JohJakob I will make those changes. I also like your idea of us

**Comment by TJScalzo** [comment]
My mouse slipped. Whoops!

**Comment by tholman** [comment]
I'm out for the rest of the week (conference) ... but looking forward to checking this out!

**Comment by TJScalzo** [comment]
I have a question for you @JohJakob. You said I should make the category in the posts normal text instead of italic. But should the category in the category page stay italic or should it also be normal text? I feel like it'd be good to stay consistent.

**Comment by JohJakob** [comment]
@TJScalzo Yes, it should be normal text, too.

**Comment by TJScalzo** [comment]
Got it @JohJakob. My text editor has bugged out for some reason and it's late so I'll make those changes tomorrow.

**Comment by TJScalzo** [comment]
Everything seems to be fixed with my editor, I don't know what happened. Anyway, I made the changes and they've been pushed. 
#### List of things I need help with:
- Colors of the categories
- Categories page (Selecting appropriate gifs)
- Neatening code (Simpler way to style things)

**Comment by TJScalzo** [comment]
I just noticed that the type WideWeirdWeb doesn't have spaces. Should I fix it so that it has spaces? I'm not the one who created that type so I'm not sure if it not having spaces was done for a specific reason. Any opinions or insight would be appreciated.

**Comment by JohJakob** [comment]
@TJScalzo In my opinion, it should have no spaces.

**Comment by TJScalzo** [comment]
@JohJakob I agree. When I tested out what it would look like with spaces it didn't look right.

**Comment by TJScalzo** [comment]
I just thought of something. The code that takes all of the information from the posts and displays it properly is duplicated on every single one of the category pages, as well as the home page. It would be more efficient to put that code into a separate file and then include that everywhere that it's needed. Should I do that or does someone else?

**Comment by JohJakob** [comment]
@TJScalzo You can do it. You know the code better than me.

**Comment by TJScalzo** [comment]
Okay, I'll do that tomorrow. I'm busy today.

**Comment by JohJakob** [comment]
@TJScalzo OK

**Comment by TJScalzo** [comment]
There are some issues to address. For example, WeirdWideWeb is so long that it overlaps with the content of the page when viewing its category page. This could possibly be solved by moving it to the left or by making it smaller. Another problem is, due to the way the posts are selected, pagination does not work on category pages. This means each category page goes on for all of the posts in that type without stopping; meaning certain pages get to be very long. I'm sure there is some way to get around this, I'll look into it.

**Comment by TJScalzo** [comment]
In my last commit I fixed the W.W.W. issue as best as I could, chose better colors, and made the categories page somewhat presentable.
#### Things left to do before this is publishable:
- Find better way to stack category buttons (Currently uses flexboxes; bad in some browsers)
- Make categories page work on mobile (It resizes strangely)

#### Optional things that'd be nice but aren't required:
- Add gifs to category buttons (I'm not sure how this should/would work. Right now they look pretty good without gifs. If someone wants to figure this out they're free to.)
- Neaten CSS (I'm almost certain things can be simplified.)
- Add pagination to category pages (Mentioned above that it currently doesn't work because of how the posts are being included/sorted.)

Anyone who has possible solutions to any of these problems, please let me know because I'm starting to struggle to find answers. ([Link to My Code](https://github.com/TJScalzo/inspiring-online))

**Comment by JohJakob** [comment]
This is super weird:

Safari:

![Safari](https://cloud.githubusercontent.com/assets/9888537/14891163/c324b5c6-0d65-11e6-9a7c-07cfdbb3ab22.png)

Google Chrome Canary:

![Google Chrome Canary](https://cloud.githubusercontent.com/assets/9888537/14891181/d14de924-0d65-11e6-935b-75e2bfedebe9.png)

**Comment by TJScalzo** [comment]
@JohJakob oh my... I have no idea what could be causing that.

**Comment by JohJakob** [comment]
@TJScalzo We shouldn't rely on Google Chrome Canary. Also, it's quite easy to do it without flexbox.

**Comment by JohJakob** [comment]
@TJScalzo I'm trying it out.

**Comment by JohJakob** [comment]
@TJScalzo Hmm, it works best with a fixed size (200×150). I would make the background colors a bit darker, so you can use white text. When hovering, the background colors could become lighter (as it already is) and the text color could change to black.

**Comment by TJScalzo** [comment]
@JohJakob so I should set the buttons to a fixed size. Is that all that needs to be done so that it works without flexboxes?

**Comment by JohJakob** [comment]
@TJScalzo I shortened the CSS to this:

```
$beauty: #D9ABFF;
$codepen: #BBEAD6;
$coding: #ABE4FF;
$madness: #FFA463;
$inspire: #FCB7A7;
$source: #FEE5AD;
$talk: #93DFB8;
$typo: #DADDFB;
$weird: #E45364;

.category-page {
  text-align: center;
}

.category-button {
  margin: 10px;
  width: 200px;
  height: 150px;
  color: #fff;
  text-align: center;
  line-height: 150px;
  border-radius: 10px;
  display: inline-block;
}

#button-beautiful {
  background: $beauty;
}

#button-codepen {
  background: $codepen;
}

#button-coding {
  background: $coding;
}

#button-css-madness {
  background: $madness;
}

#button-inspiration {
  background: $inspire;
}

#button-resource {
  background: $source;
}

#button-talk {
  background: $talk;
}

#button-typography {
  background: $typo;
}

#button-wideweirdweb {
  background: $weird;
}
```

(I added `category-page` as a class to `article`.)

**Comment by JohJakob** [comment]
@TJScalzo As you can see, it doesn't include `hover` animations.

**Comment by TJScalzo** [comment]
@JohJakob hm... The way the hover works right now is it just lowers the opacity of the entire button. We need to make it so it only changes the opacity of the background and not include the span. I think I can ignore the spans using selectors, that might work.

**Comment by TJScalzo** [comment]
[This](https://css-tricks.com/almanac/selectors/n/not/) seems like it'll work. I'm not currently at a computer. But when I get to one I'll make all of the changes @JohJakob and I have come up with.

**Comment by TJScalzo** [comment]
One question before I forget @JohJakob: Did you keep `.post` as a class for the article?

**Comment by JohJakob** [comment]
@TJScalzo Yep

**Comment by TJScalzo** [comment]
Alright! The categories are up and running! 
**Question/Idea:** Should we/I write a post that explains/introduces categories? I feel like it'd be an interesting way to throw back to the first post that @tholman wrote. It could use the same type/category that that post did, An Idea. If that post does get made it would give us a reason to add the type "An Idea" to the category list because there would then be two posts showing up in that category. Also, even if that post doesn't get made, should the "An Idea" category be added to the list of categories?

**TL;DR:** The categories work. Before pushing the changes, should a post be created explaining what was added? This would give a good reason to include the "An Idea" category on the categories page.

**Comment by JohJakob** [comment]
@TJScalzo Yeah, that's a good idea. It's a fairly big update to Inspiring Online, so we should tell the readers what's new.

**Comment by TJScalzo** [comment]
@JohJakob that was my line of thinking too. I'll make the necessary changes when I'm less busy.

**Comment by TJScalzo** [comment]
@JohJakob and anyone else: I just got an idea that I'd like feedback on. 
What if the category of a post is also listed on the home page? It wouldn't be too hard to implement and I'm sure it'd look pretty good. I'm just not sure if it's necessary and, if we decide to do it, where it should go. I was thinking underneath the title of the post in smaller text. Please let me know what you think.

**Comment by TJScalzo** [comment]
I created the announcement post. I'm pretty sure I made it way too long so I'd love it if @JohJakob, @tholman, and anyone else could read it and offer revisions.
#### Other stuff to do:
- Take screenshots of category stuff
- Turn those pictures into a gif for the post; basically an animated tour of categories (I need help with making this)
- Possibly add category labels to posts on the homepage?

**Comment by JohJakob** [comment]
@TJScalzo I'll take a look.

**Comment by TJScalzo** [comment]
I think I'm going to see how difficult it'd be to add categories to the home page post list and, if it works, show everyone how it looks.

**Comment by TJScalzo** [comment]
Well, I did it. It looks pretty good too!
![Screen Shot](https://cloud.githubusercontent.com/assets/15825214/14938658/c7f3ec18-0ef8-11e6-9442-92d45338fc00.png)
Each post has its category displayed right underneath the title. Let me know if you think there's a better place for it to go. (I couldn't think of anything other place.)

**Comment by JohJakob** [comment]
@TJScalzo Yep, that looks nice! I would increase the space between title and category a little bit.

**Comment by TJScalzo** [comment]
I've added the categories to the homepage. I also made slight changes to the Meet Categories post. I'll add a sentence soon that also says you can find the category of each post on the homepage. What I need from you @tholman and @JohJakob is help shortening and adding/removing info to/from the post. I still feel like I made it too long. Let me know what you think.

**Comment by tholman** [comment]
Open it up in a pull request, will be happy to check it out! 💃

**Comment by TJScalzo** [comment]
@tholman Okay, one second.

**Comment by JohJakob** [comment]
@TJScalzo I'm sorry, I was busy. I'll create the GIF as soon as possible.

**Comment by TJScalzo** [comment]
Don't worry @JohJakob. Whenever you get the chance I'll be happy to receive the gif. Just for reference and to make your job easy,
#### Criteria for gif:
- Zoom gif in so that the text is legible/crop it so there's not a lot of whitespace around the content of the page
- Go into a post to show category link inside of the post (AKA: Click on the title of a post and show that it has its category listed with who added the post)
- Make gif loop/repeat infinitely
- End on starting page so that it loops nicely

If your next gif has everything that the ones you've already made as well as this criteria I'm sure I'll be happy.

**Comment by JohJakob** [comment]
Take a look at [this pull request](https://github.com/TJScalzo/inspiring-online/pull/5).

**Comment by JohJakob** [comment]
@TJScalzo @Haroenv Oh, I didn't notice @Haroenv's GIF. Which one is better?

**Comment by TJScalzo** [comment]
Let me check, one second.

**Comment by Haroenv** [comment]
I think mine is more zoomed in, and smaller filesize, is that right?

**Comment by JohJakob** [comment]
@Haroenv Yeah, that's right, but it's a bit too fast in my opinion.

**Comment by TJScalzo** [comment]
I'm honestly not sure which one is better... Each one has its pluses and minuses. I think we need some outside person to decide since I can't.

**Comment by JohJakob** [comment]
@TJScalzo Wait a minute.

**Comment by JohJakob** [comment]
![Categories](https://cloud.githubusercontent.com/assets/9888537/15268033/f6829b30-19d2-11e6-9039-28f9744084bb.gif)

What about this one? It's zoomed in and only 1.6 MB at normal speed.

**Comment by Haroenv** [comment]
Yep, Thats the best one yet

**Comment by TJScalzo** [comment]
Yeah, that one seems to be the best amalgamation of the two of your's. Can you submit that as a pull request @JohJakob?

**Comment by TJScalzo** [comment]
Also, try and merge all of the changes into your fork before making the pull request @JohJakob

**Comment by JohJakob** [comment]
@TJScalzo [Here](https://github.com/TJScalzo/inspiring-online/pull/6)'s the pull request.
